### ISSUE TITLE
ghc@8.0 link error during configure

### ISSUE THREAD

**Issue by bukzor** [Linuxbrew/homebrew-core#5821]
gist-logs for my final three attempts:
  1. https://gist.github.com/anonymous/d8f190e231138abb68db90ca4c896ed5
  1. https://gist.github.com/anonymous/bd576994523d6e34a83aabecde9a3e63
  1. https://gist.github.com/anonymous/a2ee07b20bb145effede950badb661fe

`07.configure` shows the error:
```
/usr/bin/ld: /tmp/ghc@8.0-20180128-11213-dijztl/ghc-8.0.2/binary/lib/ghc-7.8.4/rts-1.0/libCffi.a(ffi64.o):
        relocation R_X86_64_32S against `.rodata' can not be used when making a shared object;
        recompile with -fPIC
/usr/bin/ld: final link failed: Nonrepresentable section on output
collect2: error: ld returned 1 exit status
configure: error: Building ghc-pwd failed
```

This reproduces reliably with `brew install ghc@8.0`. My `brew doctor` is 100% clear.
I did get a CPU lockup during one of the earlier, unlogged attempts, which may be relevant.
I ran `brew cleanup -s` before the third attempt above, with no change.

Output of brew-install, from various earlier attempts, in case it's useful: https://gist.github.com/anonymous/76110f9b830532ee7a9661f0b215e764

**Comment by bukzor** [comment]
I cleaned all state (`rm -rf $(homebrew --cache)` and `cd $(brew prefix) && git clean -fdxf`, and forced it to build from source (`export HOMEBREW_BUILD_FROM_SOURCE=1`) but the behaviour remained the same.

 * gist-log: https://gist.github.com/6e2b6e8d74d7718992339b054535193a
 * brew install output: https://gist.github.com/b3b43ab0fd908fc1936b42eaa1b11a13

**Comment by bukzor** [comment]
Apparently "PIE" is the cause, a security flag for gcc enabled by default since Ubuntu 16.10.
https://wiki.ubuntu.com/SecurityTeam/PIE

Linking objects that have different PIE setup results in this error. The ghc source ships with a binary of an older ghc, which was compiled without PIE, so their workaround is to disable PIE in that vendored, older ghc. You can find more background here: https://github.com/commercialhaskell/stack/blob/master/doc/faq.md#i-get-strange-ld-errors-about-recompiling-with--fpic

The below patch works for me, but the recommended fix (according to that FAQ) is to upgrade to GHC >= 8.0.2.

Patch: (works for me)
```!diff
--- binary/lib/ghc-7.8.4/settings.orig	2018-01-28 17:21:19.434990195 -0800
+++ binary/lib/ghc-7.8.4/settings	2018-01-28 17:21:23.419837893 -0800
@@ -1,7 +1,7 @@
 [("GCC extra via C opts", " -fwrapv"),
  ("C compiler command", "gcc"),
- ("C compiler flags", " -fno-stack-protector"),
- ("C compiler link flags", ""),
+ ("C compiler flags", "-fno-PIE -fno-stack-protector"),
+ ("C compiler link flags", "-no-pie"),
  ("Haskell CPP command","gcc"),
  ("Haskell CPP flags","-E -undef -traditional "),
  ("ld command", "ld"),
```

**Comment by bukzor** [comment]
With the above patch applied, it progressed much farther, but it still failed. I don't understand the error at all though:

```
/usr/lib/ruby/2.3.0/rubygems/basic_specification.rb:100:0: error:
    in `extensions_dir': undefined method `extension_api_version' for Gem:Module (NoMethodError)
    	from /usr/lib/ruby/2.3.0/rubygems/basic_specification.rb:91:in `extension_dir'
    	from /usr/lib/ruby/2.3.0/rubygems/basic_specification.rb:243:in `require_paths'
    	from /usr/lib/ruby/2.3.0/rubygems.rb:1116:in `register_default_spec'
    	from /usr/lib/ruby/2.3.0/rubygems/specification.rb:874:in `block in load_defaults'
    	from /usr/lib/ruby/2.3.0/rubygems/specification.rb:821:in `block in each_spec'
    	from /usr/lib/ruby/2.3.0/rubygems/specification.rb:743:in `block (2 levels) in each_gemspec'
    	from /usr/lib/ruby/2.3.0/rubygems/specification.rb:742:in `each'
    	from /usr/lib/ruby/2.3.0/rubygems/specification.rb:742:in `block in each_gemspec'
    	from /usr/lib/ruby/2.3.0/rubygems/specification.rb:741:in `each'
    	from /usr/lib/ruby/2.3.0/rubygems/specification.rb:741:in `each_gemspec'
    	from /usr/lib/ruby/2.3.0/rubygems/specification.rb:819:in `each_spec'
    	from /usr/lib/ruby/2.3.0/rubygems/specification.rb:871:in `load_defaults'
    	from /usr/lib/ruby/2.3.0/rubygems.rb:1258:in `<top (required)>'
    	from <internal:gem_prelude>:4:in `require'
    	from <internal:gem_prelude>:4:in `<internal:gem_prelude>'
`gcc' failed in phase `Assembler'. (Exit code: 1)
libraries/ghc-prim/ghc.mk:4: recipe for target 'libraries/ghc-prim/dist-install/build/GHC/Classes.o' failed
make[1]: *** [libraries/ghc-prim/dist-install/build/GHC/Classes.o] Error 1
make[1]: *** Waiting for unfinished jobs....
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
```

Full context:
https://gist.github.com/anonymous/051e869105adb6b2110fd121f6759900

**Comment by iMichka** [comment]
Regarding the patech, I do not get it. They say >= 8.0.2. We ship that version, and the patch is still needed?

**Comment by bukzor** [comment]
@iMichka I agree, that's confusing. They're all facts though.

**Comment by iMichka** [comment]
Fixed by #5996
